File: Практическая работа №3/src/main/java/io/github/ikbo0619/testing/team6/durak/Pair.java
package io.github.ikbo0619.testing.team6.durak;

import java.util.ArrayList;

public class Pair {
    private final Card attacker;
    private Card defender;

    private boolean completed;

    public Pair(Card a) {
        attacker = a;
        completed = false;
    }

    public void response(Card d) {
        setDefender(d);
        toggleCompleted();
    }

    public Card getAttacker() {
        return attacker;
    }

    public Card getDefender() {
        return defender;
    }

    public void setDefender(Card d) {
        if (isValidDefender(d)) {
            defender = d;
            return;
        }
        throw new IllegalArgumentException("Invalid defender");
    }

    public boolean isCompleted() {
        return completed;
    }

    public boolean isValidDefender(Card d) {
        try {
            if (d.trueCompareTo(attacker) > 0) {
                return true;
            }
        } catch (IllegalArgumentException e) {
            return false;
        }
        return false;
    }

    public void toggleCompleted() {
        completed = !completed;
    }

    public ArrayList<Card> fetchAllCards() {
        ArrayList<Card> ret = new ArrayList<>();
        if (completed) {
            ret.add(attacker);
            ret.add(defender);
        } else {
            ret.add(attacker);
        }
        return ret;
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder("{Pair}\n");
        for (Card card : fetchAllCards()) {
            result.append(card).append("\n");
        }
        result.append("{Pair}\n");
        return result.toString();
    }
}

File: Практическая работа №6/№3 Имеющаяся программная система/src/main/kotlin/Pair.kt
class Pair(card: Card) {
    private val attacker = card
    private var defender: Card? = null

    var completed = false
        private set

    fun response(defender: Card) {
        if (defender.trueCompareTo(attacker) > 0) {
            this.defender = defender
            completed = !completed
        }
    }

    fun fetchAllCards(): ArrayList<Card> {
        val list = ArrayList<Card>().apply {
            add(attacker)
            if (completed) {
                try { requireNotNull(defender) }
                catch (e: Exception) { println("Defender must not be null!") }

                add(defender!!)
            }
        }

        try { require(list.size <= 2) }
        catch (e: Exception) { println("List of fetched cards is invalid!") }

        return list
    }

    override fun toString() = buildString {
        append("{ ")
        for (card in fetchAllCards())
            append("$card ")
        append("}\n")
    }
}
